### PR TITLE
feat(subagent): bring Agent tool to parity with Claude Code's Agent tool

### LINKED ISSUE #70: Bring SubAgent `Agent` tool to parity with Claude Code's Agent tool
## Summary

Bring the `LmMultiTurn` sub-agent tool surface (`SubAgentToolProvider`) to full
parity with Claude Code's `Agent` tool — same interaction model, parameter
names, and description richness — and wire sub-agent templates into
`LmStreaming.Sample` for real providers (today they are test-mode only).

**Current state** (`src/LmMultiTurn/SubAgents/SubAgentToolProvider.cs`):
- `Agent` is fire-and-forget: returns the agent ID immediately; the parent must
  poll `CheckAgent`, and nothing in the description tells it to.
- One-sentence tool description; the template list is buried in the
  `template_name` parameter description with no per-template guidance.
- Parameters diverge from Claude Code: `task`/`template_name`/`agent_id` vs
  `prompt`/`subagent_type`/`name`, no `description`, no `model` override.
- Continuation is overloaded onto `Agent` via `agent_id` instead of a separate
  `SendMessage` tool.
- `LmStreaming.Sample` only enables sub-agents in `test`/`test-anthropic` modes
  (`DefaultTestAgentBuilder.CreateSubAgentOptions` returns `null`).

## Value

Models are heavily post-trained on Claude Code's Agent-tool contract
(synchronous spawn → final answer as the tool result, rich when-to-use
guidance). Matching that contract means better delegation decisions and fewer
forgotten-poll failure modes from any LLM driving our loop — and makes
sub-agents usable in the sample app with real providers instead of only in
scripted E2E tests.

## Scope

1. **Sync return model** — `Agent` blocks and returns the sub-agent's final
   answer as the tool result. Add `run_in_background: true` to opt into the
   current async behavior (returns agent ID; `CheckAgent` documents the
   polling path).
2. **Rich tool description** — add `Description`/`WhenToUse` fields to
   `SubAgentTemplate`; embed a per-template catalog and delegation guidance
   ("when to use" / "when not to") in the `Agent` tool description, mirroring
   Claude Code's format.
3. **Parameter parity** — `prompt` (was `task`), `subagent_type` (was
   `template_name`), `description` (short 3-5 word UI label), `name`, `model`
   override. Keep `add_tools`/`remove_tools` as extensions.
4. **`SendMessage` tool** — separate tool to continue an existing sub-agent by
   ID/name; remove the `agent_id` overload from `Agent`.
5. **Sample production wiring** — define sub-agent templates in
   `LmStreaming.Sample` for real providers (openai, anthropic, codex, copilot,
   claude), not just the `ITestAgentBuilder` test seam.

**Back-compat: clean break.** Rename parameters outright and update all
consumers (E2E scripted scenarios in `tests/LmStreaming.Sample.E2E.Tests`,
`tests/LmStreaming.Sample.Browser.E2E.Tests`, `tests/LmMultiTurn.Tests`) in the
same change. No aliases.

## Acceptance Criteria

- [ ] `Agent` tool call returns the sub-agent's final answer synchronously;
      `run_in_background: true` returns the agent ID immediately
- [ ] `Agent` tool description embeds the template catalog (name, description,
      when-to-use) and delegation guidance
- [ ] Parameters renamed: `prompt`, `subagent_type`, `description`, `name`,
      `model`; `SubAgentTemplate` gains `Description`/`WhenToUse`
- [ ] New `SendMessage` tool continues an existing sub-agent; `agent_id`
      removed from `Agent`
- [ ] `CheckAgent` description documents the background/polling workflow
- [ ] `LmStreaming.Sample` exposes at least one useful sub-agent template on
      real providers; sub-agent tool calls render correctly in the chat UI
- [ ] All unit + E2E tests updated and green (`LmMultiTurn.Tests`,
      `LmStreaming.Sample.E2E.Tests`, `Browser.E2E.Tests`)

**Priority:** High

### PR DESCRIPTION
**Pull request by gautam-achieveai**
## Summary

Brings the `LmMultiTurn` sub-agent tool surface (`SubAgentToolProvider`) to full
parity with Claude Code's `Agent` tool — same interaction model, parameter names,
and description richness — and wires sub-agent templates into `LmStreaming.Sample`
for real providers (previously test-mode only).

**Back-compat: clean break** — parameters renamed outright and all consumers
updated in this change. No aliases.

## What changed

1. **Sync return model.** `Agent` now blocks and returns the sub-agent's final
   answer as the tool result. `run_in_background: true` opts into async: the tool
   returns a JSON spawn receipt (agent id) immediately, `CheckAgent` polls, and the
   final result is also relayed back to the parent as a follow-up message.
2. **Rich tool description.** `SubAgentTemplate` gains `Description`/`WhenToUse`;
   the `Agent` description embeds a per-template catalog plus delegation guidance so
   the parent LLM picks the right `subagent_type`.
3. **Parameter parity.** `prompt` (was `task`), `subagent_type` (was
   `template_name`), `description` (short UI label), `name`, `model`. `add_tools`/
   `remove_tools` kept as extensions.
4. **`SendMessage` tool.** New separate tool (`target`, `prompt`,
   `run_in_background`) continues an existing sub-agent by id or name; the `agent_id`
   overload is removed from `Agent`.
5. **Sample production wiring.** `general-purpose` + `researcher` templates wired
   into the real middleware providers (OpenAI/Anthropic/Copilot) in
   `LmStreaming.Sample`. CLI providers (codex/claude/copilot) have no sub-agent hook
   and stay out of scope.

## Concurrency hardening (self-review)

The monitor releases each sub-agent's concurrency slot **exactly once per monitor**,
even when a background agent is continued in place via `SendMessage` and a single
long-lived monitor observes multiple run completions (previously this over-released
the semaphore). A synchronous caller that abandons the wait (its cancellation token
fires) now cancels the sub-agent so it stops promptly and frees its slot instead of
orphaning. Covered by a regression test that fails (agent ends `error`) if the
single-release guard is removed.

## Acceptance criteria

- [x] `Agent` returns the final answer synchronously; `run_in_background: true`
      returns the agent id immediately
- [x] `Agent` description embeds the template catalog (name, description,
      when-to-use) and delegation guidance
- [x] Parameters renamed: `prompt`, `subagent_type`, `description`, `name`,
      `model`; `SubAgentTemplate` gains `Description`/`WhenToUse`
- [x] New `SendMessage` tool continues an existing sub-agent; `agent_id` removed
      from `Agent`
- [x] `CheckAgent` documents the background/polling workflow
- [x] `LmStreaming.Sample` exposes useful sub-agent templates on real providers;
      sub-agent tool calls render in the chat UI
- [x] All unit + E2E tests updated and green

## Testing

- Solution build: 0 warnings / 0 errors (zero-warning gate)
- `LmMultiTurn.Tests`: 253 passed
- `LmStreaming.Sample.E2E.Tests`: 33 passed, 2 skipped (sandbox-gateway)
- `LmStreaming.Sample.Browser.E2E.Tests`: 13 passed, 5 skipped (CLI mock providers)

Fixes #70

### CODE REVIEW

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Deep review of PR #75 (implements #70).

## Summary

Solid architectural redesign bringing the Agent tool to Claude Code parity. The sync-by-default return model, SendMessage separation, and ReleaseGateOnce concurrency pattern are all well-executed. Test coverage is comprehensive across unit, integration, and E2E layers.

**Verdict: APPROVE** with two concurrency hardening suggestions for follow-up.

---

## Findings

### MEDIUM - Data race on `SubAgentState.Completion` (SubAgentState.cs:92-97 / SubAgentManager.cs:187,675,697)

`ResetCompletionIfFinished()` reads `Completion.Task.IsCompleted` and replaces the TCS without synchronization. The monitor thread concurrently resolves the same TCS via `TrySetResult`/`TrySetException` in `HandleRunCompletionAsync`.

**Race scenario**: Monitor resolves old TCS at the same moment `SendMessageAsync` replaces it - new TCS never resolved - synchronous caller hangs.

**Mitigation**: Currently unlikely due to monitor serialization (single async stream), but the invariant is fragile. Consider a lock or `Interlocked.CompareExchange` around the read-check-replace pattern.

### LOW - `NotifyParentOnCompletion` should be `volatile` (SubAgentState.cs:51)

Written by tool-handler thread, read by monitor thread without a memory barrier. The existing `_status` field (line 58) already uses `volatile` - apply the same for consistency and correctness.

### LOW - No timeout ceiling for synchronous Agent calls (SubAgentManager.cs:163)

`AwaitCompletionAsync` only respects the caller's `CancellationToken`. If no timeout is scoped into that token, a misbehaving sub-agent blocks the parent indefinitely. Consider documenting this contract or adding a configurable max-wait.

---

## Positive observations

- ReleaseGateOnce (lines 555-562): Correctly releases the semaphore exactly once per monitor lifetime, even across multiple RunCompletedMessages. The regression test validates this invariant directly.
- Cancellation propagation (lines 724-742): Parent cancellation promptly kills the sub-agent and releases the slot. ObjectDisposedException edge case handled.
- SendMessage as separate tool: Cleaner than the previous `agent_id` overload. Name-based addressing enables deterministic E2E test scripting.
- Template catalog: Embedding Description/WhenToUse into the Agent tool description gives the LLM sufficient context to pick the right sub-agent type. Scales fine for current template count (2).
- SubAgentExecutionException: Well-scoped, sealed, carries agent context. Used correctly in sync error path.
- Test coverage: New E2E scenarios (SendMessageTests, SubAgentBackgroundTests) + unit tests for sync/bg/error/name-resolution/gate-invariant cover the full contract.

File: src/LmMultiTurn/SubAgents/SubAgentState.cs [hunk 1]
```diff
@@ -38,6 +38,18 @@ internal class SubAgentState
...
+    /// Optional caller-supplied handle so SendMessage can address this agent by name.
+    /// </summary>
+    public string? Name { get; init; }
+
+    /// <summary>
+    /// When true (background spawn/continuation), run completion is relayed to the
+    /// parent as an injected user message. When false (synchronous call), the tool
+    /// handler awaiting <see cref="Completion"/> returns the result directly instead.
+    /// </summary>
+    public bool NotifyParentOnCompletion { get; set; }
```
Comment (gautam-achieveai) [inline src/LmMultiTurn/SubAgents/SubAgentState.cs]: **[Gautam's review bot]** LOW: This `bool` is written by `SendMessageAsync` (tool-handler thread) and read by `HandleRunCompletionAsync` (monitor thread) without a memory barrier. The JIT may cache the stale value.

The `_status` field on line 58 already uses `volatile` - apply the same treatment here:
```csharp
private volatile bool _notifyParentOnCompletion;
public bool NotifyParentOnCompletion { get => _notifyParentOnCompletion; set => _notifyParentOnCompletion = value; }
```
Non-blocking, but worth fixing for correctness/consistency.

File: src/LmMultiTurn/SubAgents/SubAgentState.cs [hunk 2]
```diff
@@ -63,4 +75,33 @@ internal class SubAgentState
...
+    public TaskCompletionSource<string> Completion { get; private set; } = CreateCompletionSource();
+
+    /// <summary>
+    /// Replaces an already-resolved completion with a fresh one so a follow-up
+    /// run (SendMessage continuation) can be awaited. A pending (unresolved)
+    /// completion is kept — existing waiters observe the next resolution.
+    /// </summary>
+    public void ResetCompletionIfFinished()
+    {
+        if (Completion.Task.IsCompleted)
```
Comment (gautam-achieveai) [inline src/LmMultiTurn/SubAgents/SubAgentState.cs]: **[Gautam's review bot]** MEDIUM: `ResetCompletionIfFinished()` reads `Completion.Task.IsCompleted` and replaces the reference without synchronization. Concurrently, the monitor thread resolves the *same* TCS via `TrySetResult`/`TrySetException`.

**Race window**: If the monitor's `HandleRunCompletionAsync` resolves the old TCS at the exact moment this method replaces it, the new TCS is never resolved and the synchronous caller hangs indefinitely.

**Currently mitigated** by the fact that the monitor is a single async stream (no true concurrency between resolution and replacement in practice), but the invariant is fragile. A lock would harden this:

```csharp
private readonly object _completionLock = new();

public void ResetCompletionIfFinished()
{
    lock (_completionLock)
    {
        if (Completion.Task.IsCompleted)
            Completion = CreateCompletionSource();
    }
}
```

And correspondingly lock around `TrySetResult`/`TrySetException` in `HandleRunCompletionAsync`.

File: src/LmMultiTurn/SubAgents/SubAgentState.cs (hunk 1 above)
Comment (gautam-achieveai) [reply src/LmMultiTurn/SubAgents/SubAgentState.cs]: [Gautam's bot] Fixed: NotifyParentOnCompletion is now a volatile-backed property mirroring the _status convention, so the tool-handler write is visible to the monitor thread without a separate barrier. Pushed in 811d4de.

File: src/LmMultiTurn/SubAgents/SubAgentState.cs (hunk 2 above)
Comment (gautam-achieveai) [reply src/LmMultiTurn/SubAgents/SubAgentState.cs]: [Gautam's bot] Fixed: added a _completionLock guarding ResetCompletionIfFinished and the completion resolution. HandleRunCompletionAsync now resolves via lock-guarded TryCompleteWithResult/TryCompleteWithException, so the read-check-replace and the resolve share one lock and can no longer race. Pushed in 811d4de.

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Incremental review of iteration 2 (811d4de).

Both findings from the first review are correctly addressed:

1. **MEDIUM (completion race)** - `_completionLock` now guards both `ResetCompletionIfFinished()` and the new `TryCompleteWithResult`/`TryCompleteWithException` methods. Since the TCS uses `RunContinuationsAsynchronously`, holding the lock during resolution won't cause continuation-in-lock issues. :white_check_mark:

2. **LOW (volatile flag)** - `NotifyParentOnCompletion` is now volatile-backed (`_notifyParentOnCompletion`), consistent with the `_status` field pattern. :white_check_mark:

The added doc comment on `AwaitCompletionAsync` clarifying the timeout story (caller's CT + MaxTurnsPerRun) is a nice touch.

**No new issues found in this iteration. LGTM — ready to merge.**
